training-service/src/services/training.py

```diff
@@ -120,15 +129,25 @@ def train(model, X_train, T_train, epochs=10):
     model.setup_for_finetuning()
     X_train, T_train = shuffle_data(X_train, T_train)
     losses = np.array([])
+    delta = 100
     for epoch in range(epochs):
-        print(f'Loading {(epoch+1/epochs)*100}%', flush=True)
+        print("Epoch: ", epoch)
         epoch_losses = np.array([])
         for idx, x in enumerate(X_train):
+            if delta < error_convergence:
+                break
             loss_of_sample = model.finetune_on(x, T_train[idx])
+            if idx != 0:
+                delta = epoch_losses[idx-1]-loss_of_sample
+                print("current delta: ", delta, flush=True)
             epoch_losses = np.append(epoch_losses, loss_of_sample)
-        avg_epoch_loss = np.mean(epoch_losses)
+        if len(epoch_losses) !=0:
+            avg_epoch_loss = np.mean(epoch_losses)
+            print(f'Average Loss {avg_epoch_loss} in epoch {epoch+1}', flush=True)  
         losses = np.append(losses, avg_epoch_loss)
-        print(f'Average Loss {avg_epoch_loss} in epoch {epoch+1}', flush=True)
+        if delta < error_convergence:
```

Comment:
`if len(epoch_losses) !=0:`is used to prevent calculating the mean on an empty array (epoch_losses). But you are right. If we switch the if statements we should not run into this problem.